### ISSUE TITLE
occasionally NullReferenceException with async query

### ISSUE THREAD

**Issue by WeihanLi** [aspnet/EntityFrameworkCore#16367]
Describe what is not working as expected.

occasionally get a NullReferenceException with async query, I'm using a `ToPagedListAsync` extension, from the exception, it's `var items = await source.Skip((pageNumber - 1) * pageSize).Take(pageSize).ToArrayAsync(cancellationToken).ConfigureAwait(false);` where exception throw

``` csharp
public static async Task<IPagedListModel<T>> ToPagedListAsync<T>(this IQueryable<T> source, int pageNumber, int pageSize, CancellationToken cancellationToken = default)
        {
            if (pageNumber <= 0)
            {
                pageNumber = 1;
            }
            if (pageSize <= 0)
            {
                pageSize = 10;
            }

            var count = await source.CountAsync(cancellationToken).ConfigureAwait(false);
            if (count == 0)
            {
                return new PagedListModel<T>() { PageNumber = pageNumber, PageSize = pageSize, TotalCount = 0 };
            }

            var items = await source.Skip((pageNumber - 1) * pageSize)
                                    .Take(pageSize)
                                    .ToArrayAsync(cancellationToken).ConfigureAwait(false);
            var pagedList = new PagedListModel<T>()
            {
                PageNumber = pageNumber,
                PageSize = pageSize,
                TotalCount = count,
                Data = items
            };

            return pagedList;
        }

```

If you are seeing an exception, include the full exceptions details (message and stack trace)，you may wanna see details on sentry <https://sentry.io/share/issue/c3e917365d974b219430ce3faef74499/>

```
Exception message:
Stack trace:
System.NullReferenceException: Object reference not set to an instance of an object. 
at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Select.cs:line 109 
at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncIterator.cs:line 98 
at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken) 

at System.Collections.Generic.AsyncEnumerableHelpers.ToArrayWithLength[T](IAsyncEnumerable`1 source, CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncEnumerableHelpers.cs:line 48
```

### Steps to reproduce
Include a complete code listing (or project/solution) that we can run to reproduce the issue.

Partial code listings, or multiple fragments of code, will slow down our response or cause us to push the issue back to you to provide code to reproduce the issue.

you can get the sample app [here](https://github.com/WeihanLi/ActivityReservation), it depend on redis service, you can run a redis service if no redis installed with the command `docker run --restart=always -d -p 6379:6379 --name redis-server redis:alpine` before run the sample app

run the app, and access the path `/api/reservation`

Exception from: <https://github.com/WeihanLi/ActivityReservation/blob/dev/ActivityReservation.API/ReservationController.cs#L46>
<https://github.com/WeihanLi/WeihanLi.EntityFramework/blob/7492a407cb4e4e6c98c324f05a36332fbda3cb9e/src/WeihanLi.EntityFramework/IQueryablePageListExtensions.cs#L77>

### Further technical details

EF Core version: 2.2.4.0/2.1.0
Database Provider: (`Pomelo.EntityFrameworkCore.MySql`)
Operating system: Docker with aspnetcore2.2
IDE: (Visual Studio 2019 16.1.4)

**Comment by smitpatel** [comment]
The problematic code path has been removed in 3.0 pipeline so this should be working. We don't know exact root cause of the issue or have a small repro code to verify the fix. I am closing this issue as fixed.

@WeihanLi - If this still fails for you, please file a new issue with repro steps and details.

**Comment by zhongzepeng** [comment]
Hello ! Any 2.XX efcore version fixed this bug，3.1 has too many breaking changes。

**Comment by ajcvickers** [comment]
@zhongzepeng There are no plans to fix this in a 2.x release. See [the planning process](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning) for info on how we decide which issues to patch.
